packages/core-data/src/hooks/use-resource-permissions.ts

```diff
@@ -110,14 +114,29 @@ type ResourcePermissionsResolution< IdType > = [
  * @template IdType
  */
 export default function useResourcePermissions< IdType = void >(
-	resource: string,
+	resource: string | EntityResource,
 	id?: IdType
```

Comment:
I am happy to do so, but my TS knowledge is lacking to do it :/

Reply:
I believe it should be doable with some [function overloads](https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads):

```typescript
function useResourcePermissions< IdType = void >(
	resource: string,
	id?: IdType
): ResourcePermissionsResolution< IdType >;

function useResourcePermissions< IdType = void >(
	resource: EntityResource,
	id?: never
): ResourcePermissionsResolution< IdType >;

function useResourcePermissions( resource, id ) {
	// ...
}

export default useResourcePermissions;
```

then, these usages should be valid:

```typescript
export function Foo( resource: string, id: number ) {
	const x = useResourcePermissions( resource, id );
	return x;
}

export function Bar( resource: EntityResource ) {
	const x = useResourcePermissions( resource );
	return x;
}
```

and this usage should be invalid, erroring that there are no overloads supporting that call:

```typescript
export function Baz( resource: EntityResource, id: number ) {
	const x = useResourcePermissions( resource, id );
	return x;
}
```

Note that in this (untested) example, we're exporting `useResourcePermissions` separately to prevent a `TS2383` error.

Let me know if that works!

Reply:
Thanks, @tyxla!

Added function overloads in [90a0d22](https://github.com/WordPress/gutenberg/pull/63653/commits/90a0d223d1d7d023100e5a70241bc9b243a8188a), and they seem to work correctly.